src/backend/distributed/executor/subplan_execution.c

```diff
@@ -66,13 +77,27 @@ ExecuteSubPlans(DistributedPlan *distributedPlan)
 		DestReceiver *copyDest = NULL;
 		ParamListInfo params = NULL;
 		EState *estate = NULL;
-
+		bool writeLocalFile = false;
 		char *resultId = GenerateResultId(planId, subPlanId);
+		List *workerNodeList =
+			FindAllWorkerNodesUsingSubplan(intermediateResultsHash,
+										   resultId);
+
+		/*
+		 * Write intermediate results to local file only if there is no worker
+		 * node that receives them
+		 */
+		if (workerNodeList == NIL)
```

Comment:
This  could happen in  two cases: (a) Subquery in  the having (b) The intermediate result is not used, such as RETURNING of a modifying CTE is not used -- for SELECT, Postgres/Citus is cleaver enough to not execute the CTE if it is not used at all, but for modifications we have to execute the queries.

So, please add a comment how this could happen

Reply:
improved the comments above.


		/*
		 * Write intermediate results to local file only if there is no worker
		 * node that receives them.
		 *
		 * This could happen in two cases:
		 * (a) Subquery in the having
		 * (b) The intermediate result is not used, such as RETURNING of a
		 *     modifying CTE is not used
		 *
		 * For SELECT, Postgres/Citus is clever enough to not execute the CTE
		 * if it is not used at all, but for modifications we have to execute
		 * the queries.
		 */